### DISCUSSION TITLE
schemas not compatible with openapi in litestar

### DISCUSSION THREAD

**Discussion by YuriFontella** [piccolo-orm/piccolo#976] [General]
I'm trying to define the correct response in the api documentation
the schema:

```py
class Jobs(Table):
    id = Serial
    name = Varchar()
    lines = Integer(default=0)
    status = Varchar(default='pending')
    uuid = UUID(default=uuid4)
    progress = Integer(default=0)
    date = Timestamptz()
```
In the route function directly assigns the class
```py
@get(path='/', summary='Retornar todos os trabalhos')
async def jobs(limit: int = 100, offset: int = 0) ->list[ Jobs]:
    ...
```
But it did not generate a response in the documentation
I tried using pydantic
```py
JobsModel = create_pydantic_model(Jobs)

@get(path='/', summary='Retornar todos os trabalhos')
async def jobs(limit: int = 100, offset: int = 0) -> list[JobsModel]:
    ...
```
generated the data in the documentation, but it didn't come with the data type, I didn't understand where the `{}` comes from

![image](https://github.com/piccolo-orm/piccolo/assets/4920619/631b1c96-70aa-4c0e-8dad-0aa549136011)
I was waiting for this result:
```
[
  {
    "name": "string",
    "lines": "integer",
    "status": "boolean",
    "uuid": "uuid",
    "progress": "integer",
    "date": "string"
  }
]
```
Does anyone know the correct way to do this?

piccolo 1.5.0
litestar 2.7.1
python 3.12.2

**Comment by dantownsend** [comment]
I think you've got a typo in your table definition, it should be:

```python
    id = Serial(primary_key=True)
```

You're right about needing to return `list[JobsModel]`, instead of `list[Jobs]`.

Which UI are you using to view the documentation? Is it the Swagger UI?

**Reply by sinisaos** [reply]
With FastAPI and BlackSheep we don't have these problems because the types are correctly generated in the schema. FastAPI example

![fastapi](https://github.com/piccolo-orm/piccolo/assets/30960668/94efce59-48e8-4e0b-9f74-f9696fa39318)

If we print `TaskModelOut.schema_json(indent=4)` Piccolo displays the correct schema with correct types. 
`{}` is shown because Litestar generates schema types eg `null or string` and should be eg `string or null`. 

![litestar](https://github.com/piccolo-orm/piccolo/assets/30960668/4a58e4cf-482b-4f46-912c-1968e81eae7a)


I even tried type hint the columns in the table definition like this, but nothing changes. 
```python
class Task(Table, db=DB):
    id: Serial = Serial(primary_key=True)
    name: Varchar = Varchar()
    completed: Boolean = Boolean(default=False)
    duration: Interval = Interval()
    task_user: ForeignKey = ForeignKey(references=BaseUser)
```

I don't think this is a Piccolo problem (Piccolo displays the correct schema with correct types), but a Litestar one. Sorry if I'm wrong.

**Reply by sinisaos** [reply]
@YuriFontella I dug a little deeper. Litestar (unlike FastAPI) uses the `oneOf` built-in generation for union types (Piccolo use Pydantic and Pydantic use `anyOf`). I don't know the reason why Litestar use `oneOf` (maybe because Litestar supports `msgspec` and `attrs` besides `Pydantic`, I don't know). If I try to change the [Litestar source code](https://github.com/litestar-org/litestar/blob/3ea7c1523de3ffb27f2dbcc609a741413f008127/litestar/_openapi/schema_generation/schema.py) to generate the schema and change all `one_of` to `any_of`, but nothing changes. If I use `all_of` (replace all `one_of` with `all_of`), Litestar correctly displays the Piccolo column types in Redoc or Swagger. 

![allof](https://github.com/piccolo-orm/piccolo/assets/30960668/d3cd1cbe-b227-488c-aa09-4c0993d6358e)

Sorry if I'm wrong but I still think that is not Piccolo problem.

**Reply by YuriFontella** [reply]
@sinisaos I can open a discussion on some litestar channel about this, since they are inegrated with piccolo.

Another test I did was generating a dataclass, and it worked showing the types.
```py
from dataclasses import dataclass

@dataclass
class JobsCustomResponse:
    name: str
    status: bool
    lines: int
    uuid: str
    progress: int
    date: str


@get(path='/', summary='Retornar todos os trabalhos')
async def jobs(limit: int = 100, offset: int = 0) -> list[JobsCustomResponse]:
    ...
```
![image](https://github.com/piccolo-orm/piccolo/assets/4920619/541c0629-469a-4904-989a-31f11d13ef22)

**Reply by sinisaos** [reply]
@YuriFontella Litestar and FastAPI seem to treat `optional` fields differently. If you try to put optional fields in a `dataclass` or a `Pydantic` model like this
```python
import typing as t
from dataclasses import dataclass
from pydantic import BaseModel

@dataclass
class JobsCustomResponse:
    name: t.Optional[str] = None
    status: t.Optional[bool] = None
    lines: t.Optional[int] = None
    uuid: t.Optional[str] = None
    progress: t.Optional[int] = None
    date: t.Optional[str] = None
# or 
class JobsCustomResponse(BaseModel):
    name: t.Optional[str] = None
    status: t.Optional[bool] = None
    lines: t.Optional[int] = None
    uuid: t.Optional[str] = None
    progress: t.Optional[int] = None
    date: t.Optional[str] = None
```
Litestar won't generate the types in the openapi docs (I don't know if this is intentional in Litestar or ?), but FastAPI will. You can do one thing and make all columns required by adding `required=True` ([docs](https://piccolo-orm.readthedocs.io/en/latest/piccolo/schema/column_types.html)) in the Piccolo table definition like this
```python
class Jobs(Table):
    id = Serial(primary_key=True, required=True)
    name = Varchar(required=True)
    lines = Integer(default=0, required=True)
    status = Varchar(default="pending", required=True)
    uuid = UUID(default=uuid.uuid4, required=True)
    progress = Integer(default=0, required=True)
    date = Timestamptz(required=True)
# create Pydantic model
JobsModel = create_pydantic_model(Jobs)
```
and then Litestar will consider them as `non-optional` (required) fields and they will be types displayed in Redoc

![required](https://github.com/piccolo-orm/piccolo/assets/30960668/ad4f7884-bbfe-4778-a5b8-43bad21166f9)

**Reply by YuriFontella** [reply]
This really worked, it's not so bad to put required in the schema fields.
Do you think this is a problem and we should notify the litestar team?

**Reply by sinisaos** [reply]
I don't know, as you wish. Regarding Litestar, I've only made one [starter template](https://github.com/sinisaos/litestar-piccolo/tree/main/backend) to use Piccolo with Litestar and I don't have enough knowledge about Litestar. Maybe you will get an answer there whether such behavior is intentional or not.

**Reply by YuriFontella** [reply]
@sinisaos thanks a lot for the help
